util/eventbus/subscribe.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package eventbus
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+)
+
+type dispatchFn func(vals *queue, acceptCh func() chan any) bool
+
+// A Queue receives events from a Bus.
+//
+// To receive events through the queue, see [Subscribe]. Subscribers
+// that share the same Queue receive events one at time, in the order
+// they were published.
+type Queue struct {
+	bus  *Bus
+	name string
+
+	write    chan any
+	stop     *goroutineShutdown
+	snapshot chan chan []any
+
+	outputsMu sync.Mutex
+	outputs   map[reflect.Type]dispatchFn
```

Comment:
It needn't be in this PR, but we should probably explain what these are for as an internal plumbing comment eventually.

Reply:
I'm not sure how I'd explain this, but you're right. How about we noodle on that in a followup PR?